packages/front-end/components/AutoGenerateMetricsModal.tsx

```diff
@@ -0,0 +1,299 @@
+import { useState, useEffect, useCallback } from "react";
+import { TrackedEventData } from "@/../back-end/src/types/Integration";
+import {
+  DataSourceInterfaceWithParams,
+  DataSourceSettings,
+} from "@/../back-end/types/datasource";
+import { cloneDeep } from "lodash";
+import { useForm } from "react-hook-form";
+import { MetricType } from "@/../back-end/types/metric";
+import track from "@/services/track";
+import { useAuth } from "@/services/auth";
+import { useDefinitions } from "@/services/DefinitionsContext";
+import Button from "./Button";
+import { DocLink } from "./DocLink";
+import Modal from "./Modal";
+import Tooltip from "./Tooltip/Tooltip";
+import AutoMetricCard from "./Settings/AutoMetricCard";
+import SelectField from "./Forms/SelectField";
+import LoadingOverlay from "./LoadingOverlay";
+
+type Props = {
+  setShowAutoGenerateMetricsModal: (show: boolean) => void;
+  datasource?: DataSourceInterfaceWithParams;
+  source: string;
+  mutate: () => void;
+};
+
+export default function AutoGenerateMetricsModal({
+  setShowAutoGenerateMetricsModal,
+  datasource,
+  source,
+  mutate,
+}: Props) {
+  const [autoMetricError, setAutoMetricError] = useState("");
+  const [trackedEvents, setTrackedEvents] = useState<TrackedEventData[]>([]);
+  const { datasources } = useDefinitions();
+  const [selectedDatasource, setSelectedDatasource] = useState<
+    DataSourceInterfaceWithParams | undefined
+  >(datasource || undefined);
+  const { apiCall } = useAuth();
+  const [loading, setLoading] = useState(false);
+
+  const form = useForm<{
+    datasourceId: string;
+    settings: DataSourceSettings | undefined;
+    metricsToCreate: {
+      name: string;
+      sql: string;
+      type: MetricType;
+    }[];
+  }>({
+    defaultValues: {
+      datasourceId: datasource?.id || "",
+      settings: datasource?.settings || {},
+      metricsToCreate: [],
+    },
+  });
+
+  const submit = async () => {
+    track("Generating Auto Metrics For User", {
+      metricsToCreate: form.watch("metricsToCreate"),
+      source,
+      type: selectedDatasource?.type,
+      dataSourceId: selectedDatasource?.id,
+      schema: selectedDatasource?.settings.schemaFormat,
+      newDatasourceForm: true,
+    });
+
+    const value = {
+      datasourceId: selectedDatasource?.id,
+      projects: selectedDatasource?.projects,
+      metricsToCreate: form.watch("metricsToCreate"),
+    };
+
+    await apiCall(`/metrics/auto-metrics`, {
+      method: "POST",
+      body: JSON.stringify(value),
+    });
+    mutate();
+  };
+
+  const getTrackedEvents = useCallback(
```

Comment:
Open to feedback on this pattern.

This form supports an optional `datasource` prop. When passed in (via the `/datasource/[did].tsx` page) we need to automatically start fetching the tracked events when the component loads. However, in the event the `datasource` prop is not passed in (when rendered via the `metrics` page) we need to fetch the tracked events only during the input field's onChange.

In order to not have duplicate code responsible for fetching the tracked events, I wrapped this in a useCallback, and I have a useEffect hook that automatically kicks off the initial fetch when loading the component, if the `datasource` prop is passed in.

This works, but I'm open to feedback on if there is a better pattern.

Reply:
You have `selectedDatasource` as state right now.  I would not do that.  Instead do something like this:

```ts
// This is derived and calculated on every render instead of stored in state
const selectedDatasource = datasource || getDatasourceById(form.watch("datasourceId"));

// Anytime the selected datasource changes, make the api call
useEffect(() => {
  // Skip if no data source is selected
  if (!selectedDatasource) return;

  // Do all of the load, apiCall, etc. logic here
}, [selectedDatasource]);
```

On first load, if a datasource is passed in as a prop, this will fetch metrics immediately.  Otherwise, it will do nothing.

Then, you can simplify the onChange handler for the data source selector.  Since `selectedDatasource` is derived instead of stored in state, you don't need to do anything other than updating the form value.

```ts
onChange={(datasourceId) => {
  form.setValue("datasourceId", datasourceId);
}}
```